pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -213,20 +242,198 @@ func (c *MCDefaultRouteController) processNextWorkItem() bool {
+		return nil
+	}
+	if !c.wireGuardInitialized {
+		var initErr error
+		defer func() {
+			if initErr != nil {
+				if err := c.cleanUpWireGuard(); err != nil {
+					klog.ErrorS(err, "Failed to clean up WireGuard")
+				}
+				c.wireGuardInitialized = false
+			}
+		}()
+		if initErr = c.initializeWireGuard(gateway); initErr != nil {
+			return initErr
+		}
+		c.wireGuardInitialized = true
+	}
+	ciImports, err := c.ciImportLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	desiredCIImports := sets.NewString()
+	var updateErr error
+	for _, ciImport := range ciImports {
+		desiredCIImports.Insert(ciImport.Name)
+		if err = c.updateSingleWireGuardPeer(ciImport); err != nil {
+			klog.V(2).ErrorS(err, "Failed to update WireGuard peer", "ClusterInfoImport", klog.KObj(ciImport))
+			updateErr = err
+		}
+	}
+	if updateErr != nil {
+		return updateErr
+	}
+
+	// Check cache and existing ClusterInfoImports, clean up routes and WireGuard peers of the
+	// removed ClusterInfoImports.
+	for ciName, ciImport := range c.installedCIImports {
+		if desiredCIImports.Has(ciName) {
+			continue
+		}
+		if err := c.removeWireGuardRouteAndPeer(ciImport); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *MCDefaultRouteController) removeWireGuardRouteAndPeer(ciImport *mcv1alpha1.ClusterInfoImport) error {
+	remoteGatewayIP, _, _ := net.ParseCIDR(ciImport.Spec.ServiceCIDR)
+	dstCIDR := net.IPNet{IP: remoteGatewayIP, Mask: net.CIDRMask(32, 32)}
+	if err := c.routeClient.DeleteRouteForLink(&dstCIDR, c.wireGuardConfig.LinkIndex); err != nil {
+		return err
+	}
+	return c.wireGuardClient.DeletePeer(ciImport.Name)
+}
+
+// updateSingleWireGuardPeer tries to update a WireGuard peer with ClusterInfoImport. If updating successfully,
+// it will also create host route to WireGuard peer.
+func (c *MCDefaultRouteController) updateSingleWireGuardPeer(ciImport *mcv1alpha1.ClusterInfoImport) error {
+	if ciImport.Spec.WireGuard == nil || ciImport.Spec.WireGuard.PublicKey == "" {
+		klog.V(2).InfoS("ClusterInfoImport's WireGuard field has not been initialized, skip it", "ClusterInfoImport", klog.KObj(ciImport))
+		return nil
+	}
+
+	klog.V(2).InfoS("Updating WireGuard peer with ClusterInfoImport", "ClusterInfoImport", klog.KObj(ciImport))
+	// The cross-cluster traffic will be both encapsulated and encrypted. To avoid routing loop, we use a tunnel endpoint
+	// IP different from the WireGuard endpoint IP. Since the ServiceCIDR is guaranteed to be unique across member clusters,
+	// we choose the ServiceCIDR's network address as the tunnel endpoint IP. For instance, if a cluster's ServiceCIDR is
+	// 10.96.0.0/16, 10.96.0.0 will be used as the tunnel endpoint IP of the cluster's Gateway Node.
+	serviceIP, serviceCIDR, err := net.ParseCIDR(ciImport.Spec.ServiceCIDR)
+	if err != nil {
+		return err
+	}
+	remoteWireGuardIP := serviceIP.Mask(serviceCIDR.Mask)
+	remoteWireGuardNet := &net.IPNet{IP: remoteWireGuardIP, Mask: net.CIDRMask(32, 32)}
```

Comment:
You mean the value is provided be user, not discovered by program?

Reply:
Can we validate the user input to avoid such input?
Anyway the calculation must be consistent in removeWireGuardRouteAndPeer and addWireGuardRouteAndPeer.

Reply:
If the users provide ServiceCIDR in `antrea-mc-config`, the controller won't try to get the ServiceCIDR of the cluster but use the CIDR users provided directly. Maybe we can add a strict validation in `antrea-mc-controller` to avoid CIDR just like `10.10.1.1/16`?